python/tvm/topi/cuda/sort.py

```diff
@@ -593,3 +694,82 @@ def schedule_topk(outs):
       The computation schedule for the op.
     """
     return _schedule_sort(outs)
+
+
+def _dyn_topk_legalize(attrs, inputs, arg_types):
+    """Legalizes dyn.topk op.
+
+    On GPU, we don't directly implement a topi kernel. Instead, we combine
+    topi sort and topi strided slice to implement topk. This topi-level composition
+    doesn't work with the dynamic op. To support the dynamic op on gpu, we instead
+    legalize it into the same logic (sort + strided slice) in relay to take advantage
+    of the VM's dynamic shape capabilities
+
+    Parameters
+    ----------
+    attrs : tvm.ir.Attrs
+        Attributes of current convolution
+    inputs : list of tvm.relay.Expr
+        The args of the Relay expr to be legalized
+    types : list of types
+        List of input and output types
+
+    Returns
+    -------
+    result : tvm.relay.Expr
+        The legalized expr
+    """
+    data = tvm.relay.var(
+        "dyn_topk_data", shape=inputs[0].checked_type.shape, dtype=inputs[0].checked_type.dtype
+    )
+    k = tvm.relay.var(
+        "dyn_topk_k", shape=inputs[1].checked_type.shape, dtype=inputs[1].checked_type.dtype
+    )
+    sort_out = tvm.relay.sort(data, axis=attrs.axis, is_ascend=attrs.is_ascend)
+    argsort_out = tvm.relay.argsort(
+        data, axis=attrs.axis, is_ascend=attrs.is_ascend, dtype=attrs.dtype
+    )
+    dshape = tvm.relay.shape_of(data)
+
+    axis = tvm.relay.const([attrs.axis])
+
+    zero = tvm.relay.const([0])
+    one = tvm.relay.const([1])
+    rank = tvm.relay.shape_of(dshape)
+
+    normalized_axis = tvm.relay.where(axis < zero, axis + rank, axis)
```

Comment:
Would these ops impact  the performance?

Reply:
Most of the small elementwise ops to get shapes will be fused, and this matches what I did in TOPI pretty closely. The only real performance concern I have with this is that it does sort and argsort separately. The way the cuda 
kernels are written, I think that ends up sorting the input data twice.